### PR TITLE
GH-44713: [Python] Add support for Decimal32 and Decimal64 types

### PR DESCRIPTION
**Pull request by paleolimbot**
### Rationale for this change

Arrow C++ and the Arrow specification now support 32-bit and 64-bit decimal types...pyarrow should too!

### What changes are included in this PR?

Added type, array, and scalar bindings.

### Are these changes tested?

Yes!

### Are there any user-facing changes?

Yes!

* GitHub Issue: #44713

### CODE REVIEW

**Comment by ianmcook** [comment]
Please add the concrete classes for the scalars to the list under the Scalars heading in `arrays.rst`.

File: python/pyarrow/__init__.py
Comment (zeroshade) [inline python/pyarrow/__init__.py:184]: You'll need to add `Decimal32Array, Decimal64Array` to line 220 and `Decimal32Scalar, Decimal64Scalar` to line 229

File: python/pyarrow/types.pxi
Comment (zeroshade) [inline python/pyarrow/types.pxi:4580]: should update to `between 1 and 18`

File: python/pyarrow/tests/test_compute.py
Comment (zeroshade) [inline python/pyarrow/tests/test_compute.py:1903]: shouldn't this be 9, not 8?

**Review by zeroshade — APPROVED** [review]

**Review by raulcd — COMMENTED** [review]
Are tests and documentation finished? Just wondering if the description of the PR has to be updated or is there something missing that is expected to be followed up on a subsequent PR

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

**Review by pitrou — APPROVED** [review]
LGTM, I just wonder if the hypothesis tests pass

**Comment by pitrou** [comment]
The CI failures are unrelated. You might want to wait for @jorisvandenbossche 's review, but you can merge as well.

**Pull request merged**
